core/store/src/flat_state.rs

```diff
@@ -92,25 +93,32 @@ mod imp {
     }
 
     #[derive(Clone)]
-    struct FlatStateCache {
+    pub struct FlatStateCache {
         // TODO: add implementation
     }
 
     impl FlatState {
+        pub fn new(
+            store: Store,
+            block_hash: CryptoHash,
+            cache: FlatStateCache,
+            flat_storage_state: FlatStorageState,
+        ) -> Self {
+            // TODO (#7327): process error gracefully
+            let blocks_to_head = flat_storage_state.get_blocks_to_head(&block_hash).unwrap();
+            Self { store, blocks_to_head, cache, flat_storage_state }
+        }
         /// Returns value reference using raw trie key, taken from the state
-        /// corresponding to `FlatState::block_hash`.
+        /// corresponding to block hash for which `FlatState` was created.
         ///
         /// To avoid duplication, we don't store values themselves in flat state,
         /// they are stored in `DBCol::State`. Also the separation is done so we
         /// could charge users for the value length before loading the value.
-        // TODO (#7327): support different roots (or block hashes).
         // TODO (#7327): consider inlining small values, so we could use only one db access.
         pub fn get_ref(&self, key: &[u8]) -> Result<Option<ValueRef>, crate::StorageError> {
-            // Take deltas ordered from `self.block_hash` to flat state head.
-            // In other words, order of deltas is the opposite of the order of blocks in chain.
-            let deltas = self.flat_storage_state.get_deltas_between_blocks(&self.block_hash)?;
-            for delta in deltas {
+            for block_hash in self.blocks_to_head.iter() {
                 // If we found a key in delta, we can return a value because it is the most recent key update.
+                let delta = self.flat_storage_state.get_delta(block_hash)?;
```

Comment:
I don't think this implementation would work because we can't afford to load deltas from storage for every `get_ref` call. that will simply be too expensive. We could afford to load deltas once for each block, but not for every `get_ref` call. 

I think here we can assume that any fork from the last final block will be at most 12 blocks long, if we implement the protocol change of forcing blocks to include empty chunks if the last final block is more than 12 blocks ago. Then in the implementation, we can always assume that all deltas for a single block can fit in memory, but then we need to deal with the case when there could be multiple forks. 

Based on that, we can modify the `deltas` field stored in FlatStorageState. Right now, it stores deltas for all blocks that are children from the final block. We will need to change that to only load new deltas when new flat state is created. We will need to implement logic to limit either the number of deltas or the total size of deltas in memory. This also likely means that we will need to change `get_trie_for_shard` to make it return error when new flat state can't be created because there are already enough deltas in memory.

Reply:
> I don't think this implementation would work because we can't afford to load deltas from storage for every get_ref call. that will simply be too expensive.

Right. The main idea here is to replace `get_delta` which getting only a single value from delta. Instead of storing the whole delta as a value for `block_hash`, I would store each value separately for tuple `(block_hash, key)`. For 12 blocks deltas will be cached, and for non-cached ones we will make N reads from deltas on disk. So for ~12s finality performance is not slowed down, and for ~100s finality performance is unlikely to drop too much, because keys for one delta are localized.

This is somewhat new idea which I also described here https://github.com/near/nearcore/issues/8436#issuecomment-1403565637 and we agreed on it on the flat storage meeting. I think it would be simpler to do for MVP. WDYT?